### ISSUE TITLE
[BUG / possible BUG] check for 3.0.8 on macOS and bash5 fails with probable known error

### ISSUE THREAD

**Issue by mcpetermohr** [drwetter/testssl.sh#2502]
I am running version 3.0.8

Is related to / couldn't find anything 

testssl.sh eas.mariagerfjord.dk

I expected the script to run without errors... But got this:

 Testing robust (perfect) forward secrecy, (P)FS -- omitting Null Authentication/Encryption, 3DES, RC4 

 PFS is offered (OK)          TLS_AES_256_GCM_SHA384 TLS_CHACHA20_POLY1305_SHA256 ECDHE-RSA-AES256-GCM-SHA384 ECDHE-RSA-AES256-SHA384
                              ECDHE-RSA-AES256-SHA ECDHE-RSA-CHACHA20-POLY1305 TLS_AES_128_GCM_SHA256 ECDHE-RSA-AES128-GCM-SHA256
                              ECDHE-RSA-AES128-SHA256 ECDHE-RSA-AES128-SHA 
 Elliptic curves offered:     prime256v1 secp384r1 secp521r1 X25519 X448 
/opt/homebrew/Cellar/testssl/3.0.8/libexec/bin/testssl.sh: line 1186: 16#: invalid integer constant (error token is "16#")
/opt/homebrew/Cellar/testssl/3.0.8/libexec/bin/testssl.sh: line 12340: 2*: syntax error: operand expected (error token is "*")

√ ~ %  ``git log | head -1``  / ``testssl.sh -v | grep from``
head: Error reading /
head: testssl.sh: No such file or directory
head: -v: No such file or directory
fatal: not a git repository (or any of the parent directories): .git
?1 ~ % git log
fatal: not a git repository (or any of the parent directories): .git

**Your system (please complete the following information):**
 - OS: macOS Sonoma
 - Platform: Darwin 23.5.0 arm64
 - OpenSSL: /opt/homebrew/opt/openssl@3/bin/openssl

**Additional context**
N/A

**Comment by drwetter** [comment]
Please provide a proper version string.

Don´t know for sure w/o looking deeper into it but the error is not knew and maybe is addressed already -- at least in the git repo.

**Comment by mcpetermohr** [comment]
Version info:

testssl.sh       3.0.8 from https://testssl.sh/

**Comment by dcooper16** [comment]
This is an issue that first came up four years ago: #2453, #2204, #1823, #1820, #1808, #1783, #1754.

However, it seems that the bug hasn't been fixed everywhere. I just tried it with the current 3.2rc3 and 3.0.8 and got the same results (see below).

Based on a quick look it seems that this is a result of a bug in testssl.sh and in the server. I think the server is sending a key_share extension with a group of 256 (ffdhe2048) but it isn't including a key. So, the key share extension has a length of 4. testssl.sh checks that the length is not less than 4, but then tries to read the length of the key (which starts at byte 5).

 I think changing testssl.sh to check that the key_share extension is at least 8 bytes would fix the issue, but I don't have time to test it at the moment. The change would be on line 14458 in the 3.2 branch and line 12370 in the 3.0 branch.

```
###########################################################
    testssl.sh       3.2rc3 from https://testssl.sh/dev/
    (5b75dd7 2024-05-29 10:55:01)

      This program is free software. Distribution and
             modification under GPLv2 permitted.
      USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!

       Please file bugs @ https://testssl.sh/bugs/

###########################################################

 Using "OpenSSL 1.0.2-bad (1.0.2k-dev)" [~179 ciphers]
 on N118978-01:./bin/openssl.Linux.x86_64
 (built: "Sep  1 14:03:44 2022", platform: "linux-x86_64")


 Start 2024-06-12 05:40:18                -->> 194.192.105.251:443 (eas.mariagerfjord.dk) <<--

 rDNS (194.192.105.251): --
 Service detected:       HTTP


 Testing robust forward secrecy (FS) -- omitting Null Authentication/Encryption, 3DES, RC4 

 FS is offered (OK)           TLS_AES_256_GCM_SHA384 TLS_CHACHA20_POLY1305_SHA256
                              ECDHE-RSA-AES256-GCM-SHA384 ECDHE-RSA-AES256-SHA384 ECDHE-RSA-AES256-SHA
                              ECDHE-RSA-CHACHA20-POLY1305 TLS_AES_128_GCM_SHA256
                              ECDHE-RSA-AES128-GCM-SHA256 ECDHE-RSA-AES128-SHA256 ECDHE-RSA-AES128-SHA
 Elliptic curves offered:     prime256v1 secp384r1 secp521r1 X25519 X448 
./testssl.sh: line 823: 16#: invalid integer constant (error token is "16#")
./testssl.sh: line 14466: 2*: syntax error: operand expected (error token is "*")
```

```
###########################################################
    testssl.sh       3.0.8 from https://testssl.sh/
    (212bdbb 2024-05-26 17:40:10)

      This program is free software. Distribution and
             modification under GPLv2 permitted.
      USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!

       Please file bugs @ https://testssl.sh/bugs/

###########################################################

 Using "OpenSSL 1.0.2-bad (1.0.2k-dev)" [~179 ciphers]
 on N118978-01:./bin/openssl.Linux.x86_64
 (built: "Sep  1 14:03:44 2022", platform: "linux-x86_64")


 Start 2024-06-12 05:41:25        -->> 194.192.105.251:443 (eas.mariagerfjord.dk) <<--

 rDNS (194.192.105.251): --
 Service detected:       HTTP


 Testing robust (perfect) forward secrecy, (P)FS -- omitting Null Authentication/Encryption, 3DES, RC4 

 PFS is offered (OK)          TLS_AES_256_GCM_SHA384 TLS_CHACHA20_POLY1305_SHA256
                              ECDHE-RSA-AES256-GCM-SHA384 ECDHE-RSA-AES256-SHA384 ECDHE-RSA-AES256-SHA
                              ECDHE-RSA-CHACHA20-POLY1305 TLS_AES_128_GCM_SHA256
                              ECDHE-RSA-AES128-GCM-SHA256 ECDHE-RSA-AES128-SHA256 ECDHE-RSA-AES128-SHA
 Elliptic curves offered:     prime256v1 secp384r1 secp521r1 X25519 X448 
./testssl.sh: line 1213: 16#: invalid integer constant (error token is "16#")
./testssl.sh: line 12378: 2*: syntax error: operand expected (error token is "*")
```

**Comment by drwetter** [comment]
Hi @dcooper16 ,

which 3.x version did you use? The one from homebrew or the released version? Asking because  the github branch has some fixes regarding those problems. Probably what's needed would be just a release. At least for me it works using macOS and Linux and: 212bdbb78e928a5441da4416eab3c7ee36dd98ef.

@mcpetermohr : that was also my (false) immediate reaction (¨fixed long time ago¨)  when seeing your issue and likely my misunderstanding, sorry.

**Comment by dcooper16** [comment]
Hi @drwetter,

I used the version from Github and tested it on Linux (bash version `5.1.16(1)-release`). So, it was https://github.com/drwetter/testssl.sh/commit/212bdbb78e928a5441da4416eab3c7ee36dd98ef for the 3.0 branch.

Did you test it against eas.mariagerfjord.dk? The reason the bug happens here but has not been seen before is that eas.mariagerfjord.dk sometimes sends a malformed key_share extension.

**Comment by drwetter** [comment]
Yes, I was testing against the same target, with the latest in both branches with both macOS and Linux.

Need to check the bash versions tomorrow.

**Comment by drwetter** [comment]
Okay, look like it happens with bash 5 version with any branch. 

My macOS shows bash 3 (surprise. not) and my Linux system yesterday is still at version 4.

**Comment by dcooper16** [comment]
I submitted PRs that make that change that I suggested yesterday. I haven't had time to thoroughly test them, but they seem to fix the problem. The change also makes sense, given that after the length check the code goes on to read the next 8 characters from `$tls_serverhello_ascii`. So, it seems appropriate to check that there are at least 8 characters available to read.

**Comment by drwetter** [comment]
Thanks a lot @dcooper16 . Worked for me. Guess I should soon make a release for 3.0.x so that brew-users get this downstream.

I didn't get though why the server key share extension seems to have an unexpected size , meaning what TLS stack is being used @mcpetermohr ?

**Comment by drwetter** [comment]
I made a new release for the 3.0 branch, looks like the homebrew system [picked it up automagically](https://formulae.brew.sh/formula/testssl#default). 

Let us know if the error persists after the update.

**Comment by mcpetermohr** [comment]
@drwetter I don't know what TLS stack the server is tunning. Some sort of TomCat server I would guess...

The issue is resolved with the 3.0.9 release in Brew :) Thanks a lot!!
